README.md
# System Dynamic Diagrams with roughjs

rough_sysdyn_diagram / You can try this [demo](https://assets.eloqueasy.de/rough_sysdyn_diagram/).

You can draw System Dynamics diagrams in rough mode. 

# Use

rough_sysdyn_diagram can be used in three modes:

* Show the diagram immediatly
* Show the diagram as a film (with delay between each "frame")
* Show the System Dynamic diagram stepwise (forward and backward) by clicking a button 

If you choose option 2 or three you can show additional informations for your audience.


## Example 1: Show diagram immediatly

![The complete diagram](screenshot_01.png "The diagram is visible immediatly")

First you have to declare a drawing area and name it in this case *svg_123*:

    <div class="row mt-3">
        <div class="col-md-9">
            <svg id="svg_123" width="800px" height="600px" viewbox="0 0 800 600"></svg>
        </div>
    </div>

Here you see the script:

    <script>
        var svg_123_model = new RoughSd("svg_123")
        model = [
           ["Level", "Mutter", 120, 240, false, true, false, false],
           ["Level", "Tochter", 300, 240, false, true, false, false],
           ["Level", "Endprod.", 480, 240, false, false, false, false]
           ["Const", "Zerfall Mutter", 180, 360, BOTTOM],
           ["Const", "Zerfall Tochter", 360, 120, TOP],

           ["Zerfall Mutter.top", "Mutter.out_rate.bottom", -5, 1],
           ["Mutter.top", "Mutter.out_rate.top", 30, 2],
           ["Zerfall Tochter.bottom", "Tochter.out_rate.top", 5, 1],
           ["Tochter.bottom", "Tochter.out_rate.bottom", -30, 2],
        ]
        svg_123_model.create_sd_model(model)
    </script>

First you create a new instance of the *RoughSd*-class with the id of your canvas (svg_123). 

    var svg_123_model = new RoughSd("svg_123")

Next you define your model as a list of objects:

Level objects use the following structure:
    
    [ 'Level', title, x, y, in-rate, out-rate, in-cloud, out-cloud]

With x and y you define the center of your symbol.

Consts objects have an easier structure:

    [ 'Const', title, x, y, position]

The position of your title can be LEFT, RIGHT, TOP or BOTTOM.

If you want to show dependencies with a link, you must declare the from- and the to-object and the bend of your link:

    [title of the from-object, title of the to-object, bend, strength]

With *.top* or *.bottom* you declare the starting or ending point of your link:

    'Zerfall Mutter.top'

If you want to connect the Rates you use *in-rate* or *out-rate*:

    'Mutter.out_rate.bottom'

After closing your model list, you call 

    svg_123_model.create_sd_model(model)
    

## Example 2: Show diagram as a film

This example hast mostly the same structure as example 1. The drawing area in this case is svg_123b, so the name
of your RoughSd-class is called *svg_123b_model*.

![The diagram as a film](screenshot_02.png "The diagram shown as a film with additional informations on the right")

The model is the same. To get a delay, you insert the number of seconds:

        model = [
           ["Level", "Mutter", 120, 240, false, true, false, false, 
                "Wir beginnen unsere Betrachtung mit einer radioaktiven Ausgangssubstanz, der sogenannten 'Mutter'-Substanz"], 2,
           ["Const", "Zerfall Mutter", 180, 360, BOTTOM, 
                "Die Zerfallskonstante der Mutter bestimmt, wie schnell dieser Zerfall stattfindet."], 2,
           ["Zerfall Mutter.top", "Mutter.out_rate.bottom", -5, 1, 
                "Die Abnahme des Materials wird mit einem stilisierten Ventil dargestellt."], 2,
           ["Mutter.top", "Mutter.out_rate.top", 30, 2, 
                "Gibt es noch viel von der Ausgangsmenge, dann kann auch viel Material zerfallen."], 2,
           ["Level", "Tochter", 300, 240, false, true, false, false, 
                "Die beim Zerfall der Mutter-Substant entstehende Substanz ist ebenfalls radioaktiv. Diese Substanz nennen wir 'Tochter'-Substanz."], 2,
           ["Const", "Zerfall Tochter", 360, 120, TOP, 
                "Wie schnell diese Abnahme erfolgt, wird durch die Zerfallskostante der Tochter-Substanz bestimmt."], 2,
           ["Zerfall Tochter.bottom", "Tochter.out_rate.top", 5, 1, 
                "Die Abnahme des Materials wird mit einem stilisierten Ventil dargestellt."], 2,
           ["Tochter.bottom", "Tochter.out_rate.bottom", -30, 2, 
                "Es zerfällt immer ein bestimmter Prozentsatz der Tochersubstanz"], 2,
           ["Level", "Endprod.", 480, 240, false, false, false, false, 
                "Am Ende des gesamten Prozesses ist ein stabiles Endprodukt entstanden."], 0
        ]

In this example every step is shown fpr 2 seconds. But you can change this delay as you want.

If you show the diagram as a film, it is a good idea to explain every step. For example you could show the 
information *Die Zerfallskonstante der Mutter bestimmt, wie schnell dieser Zerfall stattfindet* for your
Const *Zerfall Mutter*.

If you are using such informations, you have to declare a div-area with the name canvas-id *_info*: *svg_123b_info*:

    <div class="row mt-3">
        <div class="col-md-9">
            <svg id="svg_123b" width="800px" height="600px" viewbox="0 0 800 600"></svg>
        </div>
        <div class="col-md-3">
            <div class="card" style="width: 18rem;">
                <div class="card-body">
                    <h5 class="card-title">Hinweise zum Diagramm</h5>
                    <div id="svg_123b_info">Bitte auf den Rechtspfeil klicken ...</div>
                </div>
            </div>
        </div>
    </div>

To see the full example look at the file index.html

## Example 3: Show diagram stepwise

To step through the model, you have to declare two buttons: *svg_123a_btn_prev* and *svg_123a_btn_next*. If you click
on the button, you call *svg_123a_model.draw_prev_obj()* or *svg_123a_model.draw_next_obj()*:

![The diagram is shown stepwise](screenshot_03.png "The diagram is shown stepwise")

    <div class="row mt-3">
        <div class="col-md-9">
            <svg id="svg_123a" width="800px" height="600px" viewbox="0 0 800 600"></svg>
        </div>
        <div class="col-md-3">
            <div class="col-12 mb-3">
                <div class="btn-group" role="group" aria-label="forward / backward">
                    <button id="svg_123a_btn_prev" class="btn btn-light" onclick="svg_123a_model.draw_prev_obj()"><i class="fas fa-chevron-circle-left"></i></button>
                    <button id="svg_123a_btn_next" class="btn btn-light" onclick="svg_123a_model.draw_next_obj()"><i class="fas fa-chevron-circle-right"></i></button>
                </div>
            </div>
            <div class="card" style="width: 18rem;">
                <div class="card-body">
                    <h5 class="card-title">Hinweise zum Diagramm</h5>
                    <div id="svg_123a_info">Bitte auf den Rechtspfeil klicken ...</div>
                </div>
            </div>
        </div>
    </div>

The model is the same as in example 2, but you shouldn't insert delays.

Lastly you call

        svg_123a_model.create_sd_model(model, false)

With the *false*-param, the RoughSd-class "knows" to use forward- and backward-buttons.

## Example 4: A more complex example

![A more complex diagram](screenshot_04.png "A more complex diagram shown stepwise")

        model = [
           ["Level", "Beute", 240, 180, true, true, true, true, "Der große Kreislauf beginnt mit den Beutetieren."],
           ["Const", "Geb_rate Beute", 120, 60, TOP , "Die Geburtenrate bestimmt grundsätzlich, wie schnell sich die Beutetiere vermehren."],
           ["Geb_rate Beute.bottom", " Beute .in_rate.top", -15, 1, ""],
           ["Beute .top", " Beute .in_rate.top", -30, 1, "Viele Beutetiere haben auch für viel Nachwuchs. Wenige Beutetiere haben auch nur wenige Nachkommen."],
           ["Const", " Ste_rate Beute", 360, 60, TOP , "Die Sterberate beschreibt, wie lange die Tiere durchschnittlich leben."],
           ["Ste_rate Beute.bottom", " Beute .out_rate.top", 15, 1, ""],
           ["Beute .top", " Beute .out_rate.top", 30, 1, "Gibt es viele Beutetiere, so sterben natürlich auch viele Beutetiere."],
           ["Level", " Räuber ", 240, 360, true, true, true, true, "Die Welt sähe ohne Räuber ganz anders aus."],
           ["Const", " Geb_rate Räuber", 120, 480, BOTTOM , "Die Geburtenrate wirkt ähnlich wie bei den Beutetieren."],
           ["Geb_rate Räuber.top", " Räuber .in_rate.bottom", 15, 1, ""],
           ["Räuber.bottom", " Räuber .in_rate.bottom", 30, 1, "Viele Räuber haben viele Nachkommen."],
           ["Const", " Ste_rate Räuber ", 360, 480,  BOTTOM , ""],
           ["Ste_rate Räuber.top", " Räuber .out_rate.bottom", -15, 1, "Geburten- und Sterberate wirken wie bei den Beutetieren."],
           ["Räuber.bottom", " Räuber .out_rate.bottom", -30, 1, ""],
           ["Beute.bottom", " Räuber .in_rate.top", -20, 3, "Die Beutetiere sorgen dafür, dass viele Nachkommen der Räuber überleben."],
           ["Räuber.top", " Beute .out_rate.bottom", -20, 3, "Die Jäger sorgen dafür, dass die Anzahl der Beutetiere nicht zu stark anwächst."],
           ["Info", "Und damit schließt sich der Kreis."],
           ["Info", "Schön, dass du dir dieses Diagramm erarbeitet hast!"]
        ]


# Python

If you want to show the diagram immediatly, you jsut have to declare the levels, consts etc. and afterwards the connections
between theses objects. This model is saved as a normal text file, e.g. *mutter_tochter.txt*.

    mutter=Level:Mutter:4,8:out
    zerfall_mutter = Const:Zerfall Mutter|bottom:+2,+4
    tochter = Level:Tochter:10,8:out
    zerfall_tochter = Const:Zerfall Tochter|top:+2,-4
    endprodukt = Level:Endprod.:16,8:
    
    zerfall_mutter -> mutter.out:-1
    mutter.top --> mutter.out.top:6
    zerfall_tochter -> tochter.out:1
    tochter.bottom --> tochter.out.bottom:-6

Level variables are declared in the following way:

    name/id = Level : title : x,y : rates and clouds

Rates and clouds can have the following values:

    cloud_in,cloud_out
    cloud_in
    in
    cloud_out
    out

The x- and y-coordinates will bw internally multiplied with 30. You don't have to use quotes for strings.

Const object have a simpler structure:

    name/id = Const : title|position : x,y

With *position* you tell the program the position of your title. You can choose *top*, *right*, *bottom* and *left*.

As you can see in this example, the level object *mutter* has absolute x-y-values: *4,8*. Both const objects have
relative x-y-values *+2,+4*. To get the absolute psoition, these values are added to the last absolute value. In this
example: *4,8 + +2,+4 => 6,12*

To link two objects, you have to use the following form:

    from-object -> to-object : bend

If you want to strengthen the link, you can also use *-->* or *--->*. But the last doesn't look good.

*bend* will be internally multiplied by 5. With negative values you change the direction of your curve.

## Show diagram as a film or stepwise

First you declare your model with variables and links in the usual way.

    mutter=Level:Mutter:4,8:out
    zerfall_mutter = Const:Zerfall Mutter|bottom:+2,+4
    tochter = Level:Tochter:10,8:out
    zerfall_tochter = Const:Zerfall Tochter|top:+2,-4
    endprodukt = Level:Endprod.:16,8:
    
    zerfall_mutter -> mutter.out:-1
    mutter.top --> mutter.out.top:6
    zerfall_tochter -> tochter.out:1
    tochter.bottom --> tochter.out.bottom:-6
    
    mutter (2) = Wir beginnen unsere Betrachtung mit einer radioaktiven Ausgangssubstanz, der sogenannten "Mutter"-Substanz
    zerfall_mutter (2) = Die Zerfallskonstante der Mutter bestimmt, wie schnell dieser Zerfall stattfindet.
    zerfall_mutter --> mutter.out (2) = Die Abnahme des Materials wird mit einem stilisierten Ventil dargestellt.
    mutter -> mutter.out (2) = Gibt es noch viel von der Ausgangsmenge, dann kann auch viel Material zerfallen.
    tochter (2) = Die beim Zerfall der Mutter-Substant entstehende Substanz ist ebenfalls radioaktiv. Diese Substanz nennen wir "Tochter"-Substanz.
    zerfall_tochter (2) = Wie schnell diese Abnahme erfolgt, wird durch die Zerfallskostante der Tochter-Substanz bestimmt.
    zerfall_tochter -> tochter.out (2) = Die Abnahme des Materials wird mit einem stilisierten Ventil dargestellt.
    tochter -> tochter.out (2) = Es zerfällt immer ein bestimmter Prozentsatz der Tochersubstanz
    endprodukt = Am Ende des gesamten Prozesses ist ein stabiles Endprodukt entstanden.

Then you have to declare the sequence by naming the object name and/or the link. If you want to show a film, then you
have to tell the program the time in seconds. Example *(2)*. Then comes the equal sign and you can give more informations
for this object or link:

    name/id (delay) = informations
    from_object -> to_object (delay) = informations

If you want to show the diagram stepwise, you renove the delay values.

## More complex example

The model:

       ["Level", "Beute", 240, 180, true, true, true, true, "Der große Kreislauf beginnt mit den Beutetieren."],
       ["Const", "Geb_rate Beute", 120, 60, TOP , "Die Geburtenrate bestimmt grundsätzlich, wie schnell sich die Beutetiere vermehren."],
       ["Geb_rate Beute.bottom", "Beute.in_rate.top", -15, 1, ""],
       ["Beute .top", "Beute.in_rate.top", -30, 1, "Viele Beutetiere haben auch für viel Nachwuchs. Wenige Beutetiere haben auch nur wenige Nachkommen."],
       ["Const", "Ste_rate Beute", 360, 60, TOP , "Die Sterberate beschreibt, wie lange die Tiere durchschnittlich leben."],
       ["Ste_rate Beute.bottom", "Beute.out_rate.top", 15, 1, ""],
       ["Beute.top", "Beute.out_rate.top", 30, 1, "Gibt es viele Beutetiere, so sterben natürlich auch viele Beutetiere."],
       ["Level", "Räuber", 240, 360, true, true, true, true, "Die Welt sähe ohne Räuber ganz anders aus."],
       ["Const", "Geb_rate Räuber", 120, 480, BOTTOM , "Die Geburtenrate wirkt ähnlich wie bei den Beutetieren."],
       ["Geb_rate Räuber.top", "Räuber.in_rate.bottom", 15, 1, ""],
       ["Räuber.bottom", "Räuber.in_rate.bottom", 30, 1, "Viele Räuber haben viele Nachkommen."],
       ["Const", "Ste_rate Räuber", 360, 480,  BOTTOM , ""],
       ["Ste_rate Räuber .top", "Räuber.out_rate.bottom", -15, 1, "Geburten- und Sterberate wirken wie bei den Beutetieren."],
       ["Räuber.bottom", " Räuber.out_rate.bottom", -30, 1, ""],
       ["Beute.bottom", " Räuber.in_rate.top", -20, 3, "Die Beutetiere sorgen dafür, dass viele Nachkommen der Räuber überleben."],
       ["Räuber.top", " Beute.out_rate.bottom", -20, 3, "Die Jäger sorgen dafür, dass die Anzahl der Beutetiere nicht zu stark anwächst."],
       ["Info", "Und damit schließt sich der Kreis."],
       ["Info", "Schön, dass du dir dieses Diagramm erarbeitet hast!"]

If you just want to show a text information, you can use *Info*.

## Use

    python generate_sysdyn_diagram.py filename

# See also

* [Roughjs](https://roughjs.com/)


CHANGES.txt
sysdyn_diagramm.js

0.83 2022-09-21 URL_SNS inside class; some trims
                generate_sysdyn_diagram.py: some strips()
0.82 2022-09-20 rough_sysdyn_diagram.js 
0.81 2022-09-11 ready to publish
0.51 2022-09-10 show as a film / show stepwise
0.50 2022-09-09 prototype
0.05 2022-09-06 begin


generate_sysdyn_diagram.py
"""
generate_sys_dyn_diagram.py

Generates system dynamics diagramm in HTML format based on input file

Hans Rauch
(c) MIT LICENSE
2022-09-15 / 2022-09-20
Version 0.82
"""

import random
import re
import sys
from easydict import EasyDict as edict

model_vars = edict({})
connections = edict({})
sequence = []


def get_model_data(obj: dict, info: str) -> str:
    """
    Return the HTML-code for <obj> and <info>
    """

    def get_js(value: bool) -> str:
        """
        Return Javascript boolean
        """
        if value:
            return 'true'
        else:
            return 'false'

    line = f'           ["{obj.obj}", "{obj.title}", {30 * obj.x}, {30 * obj.y}'
    if obj.obj == 'Level':
        line += f', {get_js(obj.cloud_in)}, {get_js(obj.rate_out)}, {get_js(obj.cloud_in)}, {get_js(obj.cloud_out)}'
    elif obj.obj == 'Const':
        line += f', {obj.position}'
    else:
        assert False
    line_1 = line + f', "{info}"]'
    line += ']'
    return line, line_1


def get_connection_data(conn: dict, info: str)-> str:
    """
    Return HTML-code for <connection> and <info>
    """

    def get_name_position(obj: str) -> {str, str}:
        """
        Returns the name and the position of title
        """
        n_pos = obj.rfind('.')
        position = obj[n_pos+1:]
        name = obj[:n_pos]
        n_pos = name.rfind('.')
        if n_pos > 0:
            position = f'{name[n_pos+1:]}.{position}'
        return name, position

    def get_raw_name(name: str) -> str:
        """
        Returns the raw name of <name>
        """
        n_pos = name.find('.')
        if n_pos > 0:
            return name[:n_pos]
        else:
            return name

    conn.obj_from = conn.obj_from.replace('.in.', '.in_rate.').replace('.out.', '.out_rate.')
    conn.obj_to = conn.obj_to.replace('.in.', '.in_rate.').replace('.out.', '.out_rate.')
    from_name, from_position = get_name_position(conn.obj_from)
    to_name, to_position = get_name_position(conn.obj_to)
    m_from = model_vars[get_raw_name(from_name)]
    m_to = model_vars[get_raw_name(to_name)]
    line = f'           ["{m_from.title}.{from_position}", "{m_to.title}.{to_position}", {5 * conn.bend}, {conn.strength}'
    line_1 = line + f', "{info}"]'
    line += ']'
    return line, line_1


def get_info_data(delay: int, info: str) -> str:
    """
    Return the HTML-code for <delay> and <info>
    """
    line = f'           ["Info", "{info}"]'
    return line, line


def generate_html(sequence: list) -> str:
    """
    Creates HTML block
    """
    UID = 'r' + str(random.random())[2:11]
    model_0 = ''
    model_1 = ''
    model_2 = ''

    for obj in sequence:
        id = obj[0]
        delay = obj[1]
        info = obj[2]
        if id in model_vars:
            line_0, line_1 = get_model_data(model_vars[id], info)
        elif id in connections:
            line_0, line_1 = get_connection_data(connections[id], info)
        else:
            line_0, line_1 = get_info_data(delay, info)
        model_0 += f'{line_0},\n'
        model_1 += f'{line_1}, {delay},\n'
        model_2 += f'{line_1},\n'

    # HTML template
    data = f"""
        <!-- immediatly -->
        <div class="row mt-3">
            <div class="col-md-9">
                <svg id="svg_{UID}0" width="800px" height="600px" viewbox="0 0 800 600"></svg>
            </div>
        </div>
        <script>
            var svg_{UID}0_model = new RoughSd("svg_{UID}0")
            model = [
{model_0[:-2]}
            ]
            svg_{UID}0_model.create_sd_model(model)
        </script>

        ------------------------------------------------------

        <!--  as a film -->
        <div class="row mt-3">
            <div class="col-md-9">
                <svg id="svg_{UID}1" width="800px" height="600px" viewbox="0 0 800 600"></svg>
            </div>
            <div class="col-md-3">
                <div class="card" style="width: 18rem;">
                    <div class="card-body">
                        <h5 class="card-title">Hinweise zum Diagramm</h5>
                        <div id="svg_{UID}1_info">Bitte auf den Rechtspfeil klicken ...</div>
                    </div>
                </div>
            </div>
        </div>
        <script>
            var svg_{UID}_model = new RoughSd("svg_{UID}1")
            model = [
{model_1[:-2]}
            ]
            svg_{UID}1_model.create_sd_model(model)
        </script>

        -----------------------------------------------------------------

        <!-- stepwise -->
        <div class="row mt-3">
            <div class="col-md-9">
                <svg id="svg_{UID}2" width="800px" height="600px" viewbox="0 0 800 600"></svg>
            </div>
            <div class="col-md-3">
                <div class="col-12 mb-3">
                    <div class="btn-group" role="group" aria-label="forward / backward">
                        <button id="svg_{UID}2_btn_prev" class="btn btn-light" onclick="svg_{UID}2_model.draw_prev_obj()"><i class="fas fa-chevron-circle-left"></i></button>
                        <button id="svg_{UID}2_btn_next" class="btn btn-light" onclick="svg_{UID}2_model.draw_next_obj()"><i class="fas fa-chevron-circle-right"></i></button>
                    </div>
                </div>
                <div class="card" style="width: 18rem;">
                    <div class="card-body">
                        <h5 class="card-title">Hinweise zum Diagramm</h5>
                        <div id="svg_{UID}_info">Bitte auf den Rechtspfeil klicken ...</div>
                    </div>
                </div>
            </div>
        </div>
        <script>
            var svg_{UID}2_model = new RoughSd("svg_{UID}2")
            model = [
{model_2[:-2]}
            ]
            svg_{UID}2_model.create_sd_model(model, false)
        </script>
    """
    return data


def get_var_name(name: str) -> str:
    """
    Extracts the variable name
    """
    if name.lower() == 'true':
        name = 'true'
    elif name.lower() == 'false':
        name = 'false'
    elif name.lower() == 'none':
        name = 'null'
    return name


def generate_sd_diagram (model_raw):
    """
    Checks model_raw and creates HTML-page
    """

    def get_js_string(s: str) ->str:
        """
        Removes '"' from string
        """
        return s.strip().replace('"', "'")


    def get_name(line: str) -> {str, list}:
        """
        """
        n_pos = line.find('=')
        name = line[:n_pos].strip().lower()
        return name, line[n_pos+1:].split(':')

    def get_var(line: str, abs_x: float, abs_y: float) -> {float, float}:
        """
        Extracts variable
        """
        name, attrs = get_name(line)
        name = name.strip()
        obj = attrs[0].strip()
        s = attrs[1].split('|')
        title = s[0].strip()
        if len(s) == 1:
            position = 'TOP'
        else:
            position = s[1].upper()
        xy = attrs[2].split(',')
        if attrs[2].find('+') >=0 or attrs[2].find('-') >= 0:
            x = abs_x + eval(xy[0])
            y = abs_y + eval(xy[1])
        else:
            x = eval(xy[0])
            y = eval(xy[1])
            abs_x = x
            abs_y = y
        cloud_in = False
        rate_in = False
        rate_out = False
        cloud_out = False
        if obj == 'Level':
            in_out = attrs[3].split(',')
            if len(in_out) == 2:
                if in_out[0].find('in') >= 0:
                    rate_in = True
                if in_out[0].find('cloud') >= 0:
                    cloud_in = True
                if in_out[1].find('out') >= 0:
                    rate_out = True
                if in_out[1].find('cloud') >= 0:
                    cloud_out = True
            else:
                in_out = in_out[0]
                if in_out.find('in') >= 0:
                    rate_in = True
                    if in_out.find('cloud') >= 0:
                        cloud_in = True
                elif in_out.find('out') >= 0:
                    rate_out = True
                    if in_out.find('cloud') >= 0:
                        cloud_out = True
        if not name in model_vars:
            if obj == 'Level':
                model_vars[name] = edict({
                    'id': name,
                    'obj': obj,
                    'title': title,
                    'x': x,
                    'y': y,
                    'cloud_in': cloud_in,
                    'rate_in': rate_in,
                    'rate_out': rate_out,
                    'cloud_out': cloud_out
                })
            else:
                model_vars[name] = edict({
                    'id': name,
                    'obj': obj,
                    'title': title,
                    'position': position,
                    'x': x,
                    'y': y
                })
        else:
            node = model_vars[name]
            if 'delay' in node and 'info' in node:
                sequence.append((name, node.delay, get_js_string(node.info)))
            elif 'delay' in node:
                sequence.append((name, node.delay, ''))
            elif 'info' in node:
                sequence.append((name, 0, get_js_string(node.info)))
            else:
                sequence.append((name, 0, ''))
        return abs_x, abs_y


    def get_info(line: str):
        """
        Extracts additional informations
        """
        name, attrs = get_name(line)
        n_pos = name.find('(')
        if n_pos > 0:
            delay = eval(name[n_pos+1:].replace(')', ''))
            name = name[:n_pos].strip()
        else:
            delay = 0
        model_vars[name]['info'] = attrs[0].strip()
        model_vars[name]['delay'] = delay
        return name, delay, attrs[0]


    def get_y(node: str) -> int:
        """
        Get y-position of node
        """
        n_pos = node.find('.')
        if n_pos > 0:
            node = node[:n_pos]
        return model_vars[node].y


    def get_position(conn: str, position: str) -> {str, str}:
        """
        Gets the position of this connection
        """
        n_pos = conn.rfind('.')
        if n_pos > 0:
            last = conn[n_pos+1:]
            if last in ['top', 'bottom', 'left', 'right']:
                position = last
                conn = conn[:n_pos]
        return conn, position


    def add_connection(conn_from: str, conn_to: str, bend: int, strength: int=1):
        """
        Adds information for this connection
        """
        position = ''
        conn_from = conn_from.strip()
        conn_to = conn_to.strip()
        conn_from, position_from = get_position(conn_from, position)
        conn_to, position_to = get_position(conn_to, position)

        if conn_from in ['in', 'out']:
            conn_from = f'{conn_to}.{conn_from}'
        if conn_to in ['in', 'out']:
            conn_to = f'{conn_from}.{conn_to}'
        id = f'{conn_from}-{conn_to}'
        if not position_from and not position_to:
            y_from = get_y(conn_from)
            y_to = get_y(conn_to)
            if y_from < y_to:
                conn_from += '.bottom'
                conn_to += '.top'
            else:
                conn_from += '.top'
                conn_to += '.bottom'
        elif not position_from:
            conn_from += f'.{position_to}'
        elif not position_to:
            conn_to += f'.{position_from}'
        if position_from:
            conn_from += f'.{position_from}'
        if position_to:
            conn_to += f'.{position_to}'
        if not id in connections:
            connections[id] = edict({
                'id': id,
                'obj_from': conn_from,
                'obj_to': conn_to,
                'strength': strength,
                'bend': bend
            })
        else:
            sequence.append((id, 0, ''))

    def get_conn_info(line: str):
        """
        Extracts additional informations
        """
        name, attrs = get_name(line)
        n_pos = name.find('(')
        if n_pos > 0:
            delay = eval(name[n_pos+1:].replace(')', ''))
            name = name[:n_pos].strip()
        else:
            delay = 0
        return name, delay


    def get_connection(line: str):
        """
        Gets the connection declared in <line>
        """
        m = re.search('-*>', line)
        if m:
            s_right_arrow = line[m.start():m.end()]
            conn_from, conn_to = line.split(s_right_arrow)
            arr = conn_to.split(':')
            if len(arr) == 1:
                bend = 5
            else:
                conn_to = arr[0]
                bend = eval(arr[1])
            add_connection(conn_from, conn_to, bend, m.end() - m.start() - 1)
        elif line in model_vars or line in connections:
            sequence.append((line, 0, ''))
        else:
            m = re.search('\(\d*\)', line)
            if m:
                delay = eval(line[m.start():m.end()])
                info = line[m.end():].strip()
            else:
                delay = 0
                info = line.strip()
            sequence.append(('', delay, get_js_string(info)))


    abs_x = -1
    abs_y = -1
    lines = model_raw.split('\n')
    for line in lines:
        # Remove comments
        n_comment = line.find('#')
        if n_comment >= 0:
            line = line[:n_comment]
        line = line.strip()
        if line:
            n_pos = line.find('=')
            if n_pos > 0:
                m = re.search('-*>', line)
                if line.find(':') > n_pos:
                    # declaration of variable
                    abs_x, abs_y = get_var(line, abs_x, abs_y)
                elif not m:
                    # information for variable
                    id, delay, info  = get_info(line)
                    sequence.append((id, delay, get_js_string(info)))
                else:
                    # information for connection
                    info = line[n_pos+1:].strip()
                    line = line[:n_pos]
                    conn_from = line[:m.start()].strip()
                    line = line[m.end():]
                    name, delay = get_conn_info(line)
                    if name in['in', 'out']:
                        conn_to = f'{conn_from}.{name}'
                    else:
                        conn_to = name
                    id = f'{conn_from}-{conn_to}'
                    sequence.append((id, delay, get_js_string(info)))
            else:
                # connections
                get_connection(line)
    if not sequence:
        for name in model_vars:
            sequence.append((name, 0, ''))
        for name in connections:
            sequence.append((name, 0, ''))
    return generate_html(sequence)

# read input file and generate output file
if __name__ == "__main__":
    if len(sys.argv) < 2:
        print ('Usage: python generate_sys_dyn_diagram.py <filename>')
    else:
        filename = sys.argv[1]
        n_pos = filename.rfind('.')
        if n_pos > 0:
            f_name = filename[:n_pos]
        else:
            f_name = filename
        f_name += '.html'

        with open(filename, 'r') as f:
            model_raw = f.read()
        sd_diagram_text = generate_sd_diagram(model_raw)
        with open(f_name, 'w') as f:
            f.write(sd_diagram_text)
        print (f'Result: {f_name}')
